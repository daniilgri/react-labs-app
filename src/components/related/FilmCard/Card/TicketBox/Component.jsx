import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { Redirect, withRouter } from "react-router-dom";

import {
  Wrapper,
  Title,
  Content,
  OptionList,
  OptionTitle,
  ButtonOption,
  OrderButton,
  Box,
  TimesList,
  TimeOptionButton,
  DateValue,
  Loading,
} from "./styles";

const Component = ({
  screeningDates,
  user,
  onOrder,
  orderLoading,
  orderError,
  history,
}) => {
  const [chosenDate, setChosenDate] = useState({ date: "", time: "" });

  const handleOrderButtonOnClick = (event) => {
    event.preventDefault();
    if (user) {
      if (chosenDate.date !== "") {
        onOrder({ chosenDate });
        history.push("/profile/orders");
      }
    } else {
      return <Redirect to="/login" />;
    }
  };

  const handleTimeOptionButtonOnClick = (sd) => (event) => {
    setChosenDate(sd);
  };

  const listTimes = (dateItem) => {
    return (timeItem) => (
      <TimeOptionButton
        chosen={chosenDate.time === timeItem}
        key={uuidv4()}
        onClick={handleTimeOptionButtonOnClick({
          date: dateItem.date,
          time: timeItem,
        })}
      >
        {timeItem}
      </TimeOptionButton>
    );
  };

  useEffect(() => {}, [orderLoading]);

  return (
    <Wrapper>
      <Title>Ticketing</Title>
      <Content>
        <Box>
          <OptionTitle>Day</OptionTitle>
          <OptionList>
            {screeningDates.length > 0 &&
              screeningDates.map((dateItem) => (
                <ButtonOption key={uuidv4()}>
                  <DateValue chosen={chosenDate.date === dateItem.date}>
                    {dateItem.date}
                  </DateValue>
                  <TimesList>
                    {dateItem.times.map(listTimes(dateItem))}
                  </TimesList>
                </ButtonOption>
              ))}
          </OptionList>
        </Box>
      </Content>
      {orderLoading && <Loading>Loading...</Loading>}
      <OrderButton type="button" onClick={handleOrderButtonOnClick}>
        Order
      </OrderButton>
    </Wrapper>
  );
};

Component.defaultProps = {
  screeningDates: [],
  user: null,
  onOrder: () => {},
  orderLoading: false,
  orderError: "",
};

Component.propTypes = {
  screeningDates: PropTypes.array.isRequired,
  user: PropTypes.object,
  onOrder: PropTypes.func.isRequired,
  orderLoading: PropTypes.bool.isRequired,
  orderError: PropTypes.string,
};

export default withRouter(Component);
