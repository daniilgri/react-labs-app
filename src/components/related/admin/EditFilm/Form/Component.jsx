import React, { useEffect } from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useFormik } from "formik";

import DescriptionSection from "./DescriptionSection";
import MediaSection from "./MediaSection";
import FormalSection from "./FormalSection";

import { Wrapper, Head, Title, Body, FilledButton, Loading, RedirectLink } from "./styles";

const Component = ({
  openAddScreeningDateModal,
  filmId,
  film,
  editFilmRequested,
  fetchEditFilmRequested,
  loading,
  error,
}) => {
  useEffect(() => {
    fetchEditFilmRequested(filmId);
  }, [fetchEditFilmRequested, filmId]);

  const formik = useFormik({
    initialValues: {
      title: film.title,
      description: film.description,
      ticketPrice: film.ticketPrice,
      imageAsFile: film.image,
      previewImage: film.image,
      tags: film.tags,
      screeningDates: film.screeningDates,
    },
    onSubmit: values => {
      const changedValues = {};
      Object.keys(values).forEach(prop => {
        if (formik.initialValues[prop] !== values[prop]) {
          changedValues[prop] = values[prop];
        }
      });
      if (Object.keys(changedValues).length !== 0 && changedValues.constructor === Object) {
        editFilmRequested({ values, changedValues, filmId: film.id });
      }
    },
    enableReinitialize: true,
  });

  const handleImageUpload = event => {
    event.preventDefault();
    formik.setFieldValue("imageAsFile", event.currentTarget.files[0]);
  };

  const handleScreeningDateAddition = screeningDate => {
    formik.setFieldValue("screeningDates", [...formik.values.screeningDates, screeningDate]);
  };

  const handleScreeningDateDelete = screeningDate => {
    formik.setFieldValue(
      "screeningDates",
      formik.values.screeningDates.filter(sd => sd.date !== screeningDate.date)
    );
  };

  const handleTagAddition = tag => {
    formik.setFieldValue("tags", [...formik.values.tags, tag]);
  };

  const handleTagDelete = tag => {
    formik.setFieldValue(
      "tags",
      formik.values.tags.filter(elTag => elTag !== tag)
    );
  };

  const handleFormOnKeyDown = event => {
    if (event.key === "Enter") {
      event.preventDefault();
    }
  };

  if (loading || error) {
    return <Loading>Loading...</Loading>;
  }

  return (
    <Wrapper onKeyDown={handleFormOnKeyDown} onSubmit={formik.handleSubmit}>
      <Head>
        <Title>Edit film - {film.title}</Title>{" "}
        <RedirectLink to={`/admin/film/${film.id}`}>Go to film card</RedirectLink>
      </Head>
      <Body>
        <DescriptionSection
          title={formik.values.title}
          description={formik.values.description}
          tags={formik.values.tags}
          errors={formik.errors}
          onChange={formik.handleChange}
          onTagAdd={handleTagAddition}
          onTagDelete={handleTagDelete}
        />
        <MediaSection
          errors={formik.errors}
          onImageUpload={handleImageUpload}
          previewImage={formik.values.previewImage}
        />
        <FormalSection
          screeningDates={formik.values.screeningDates}
          ticketPrice={formik.values.ticketPrice}
          errors={formik.errors}
          onChange={formik.handleChange}
          onAddScreeningDateModalOpen={openAddScreeningDateModal}
          onScreeningDateAdd={handleScreeningDateAddition}
          onScreeningDateDelete={handleScreeningDateDelete}
        />
      </Body>

      <FilledButton type="submit">Publish</FilledButton>
    </Wrapper>
  );
};

Component.defaultProps = {
  error: "",
  filmId: "",
};

Component.propTypes = {
  editFilmRequested: PropTypes.func.isRequired,
  openAddScreeningDateModal: PropTypes.func.isRequired,
  filmId: PropTypes.string,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ).isRequired,
  fetchEditFilmRequested: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
};

export default withRouter(Component);
