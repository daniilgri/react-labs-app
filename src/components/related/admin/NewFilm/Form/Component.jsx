import React from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useFormik } from "formik";

import DescriptionSection from "./DescriptionSection";
import MediaSection from "./MediaSection";
import FormalSection from "./FormalSection";

import newFilmSchema from "../../../../../validations/newFilmSchema";
import { Wrapper, Head, Title, Body, FilledButton } from "./styles";

const Component = ({ addFilmRequested, history, openAddScreeningDateModal }) => {
  const formik = useFormik({
    initialValues: {
      title: "",
      description: "",
      ticketPrice: 0,
      imageAsFile: "",
      tags: [],
      screeningDates: [],
    },
    validationSchema: newFilmSchema,
    onSubmit: values => {
      addFilmRequested(values);
      history.push("/");
    },
  });

  const handleImageUpload = event => {
    event.preventDefault();
    formik.setFieldValue("imageAsFile", event.currentTarget.files[0]);
  };

  const handleScreeningDateAddition = screeningDate => {
    formik.setFieldValue("screeningDates", [...formik.values.screeningDates, screeningDate]);
  };

  const handleScreeningDateDelete = screeningDate => {
    formik.setFieldValue(
      "screeningDates",
      formik.values.screeningDates.filter(sd => sd.date !== screeningDate.date)
    );
  };

  const handleTagAddition = tag => {
    formik.setFieldValue("tags", [...formik.values.tags, tag]);
  };

  const handleTagDelete = tag => {
    formik.setFieldValue(
      "tags",
      formik.values.tags.filter(elTag => elTag !== tag)
    );
  };

  const handleFormOnKeyDown = event => {
    if (event.key === "Enter") {
      event.preventDefault();
    }
  };

  return (
    <Wrapper onKeyDown={handleFormOnKeyDown} onSubmit={formik.handleSubmit}>
      <Head>
        <Title>Add new film</Title>
      </Head>
      <Body>
        <DescriptionSection
          values={formik.values}
          errors={formik.errors}
          onChange={formik.handleChange}
          onTagAdd={handleTagAddition}
          onTagDelete={handleTagDelete}
        />
        <MediaSection errors={formik.errors} onImageUpload={handleImageUpload} />
        <FormalSection
          values={formik.values}
          errors={formik.errors}
          onChange={formik.handleChange}
          onAddScreeningDateModalOpen={openAddScreeningDateModal}
          onScreeningDateAdd={handleScreeningDateAddition}
          onScreeningDateDelete={handleScreeningDateDelete}
        />
      </Body>

      <FilledButton type="submit">Publish</FilledButton>
    </Wrapper>
  );
};

Component.propTypes = {
  addFilmRequested: PropTypes.func.isRequired,
  history: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.func, PropTypes.number, PropTypes.string, PropTypes.object])
  ).isRequired,
  openAddScreeningDateModal: PropTypes.func.isRequired,
};

export default withRouter(Component);
