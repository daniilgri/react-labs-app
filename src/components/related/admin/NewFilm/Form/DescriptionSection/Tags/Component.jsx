import React, { useState } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";

import { Wrapper, List, Item, TagButton, Input, ErrorText } from "./styles";

const Component = ({ onSet, onDelete, values, errors }) => {
  const [tagInput, setTagInput] = useState("");

  const handleInputOnKeyDown = event => {
    if (event.key === "Enter") {
      if (!values.includes(tagInput) && !(tagInput === "")) {
        onSet(tagInput);
        setTagInput("");
      }
    }
  };

  const handleDeleteTagButtonOnClick = tag => () => {
    onDelete(tag);
  };

  const handleInputOnChange = event => {
    setTagInput(event.target.value);
  };

  return (
    <Wrapper>
      <List>
        {values.length > 0
          ? values.map(el => (
              // eslint-disable-next-line react/jsx-indent
              <Item key={uuidv4()}>
                {el}
                <TagButton type="button" onClick={handleDeleteTagButtonOnClick(el)}>
                  -
                </TagButton>
              </Item>
            ))
          : null}
      </List>
      <Input
        onKeyDown={handleInputOnKeyDown}
        placeholder="Type tag"
        name="tagInput"
        id="tagInput"
        value={tagInput}
        onChange={handleInputOnChange}
      />
      {errors.tags && <ErrorText>{errors.tags}</ErrorText>}
    </Wrapper>
  );
};

Component.propTypes = {
  onSet: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  values: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  errors: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
};

export default Component;
