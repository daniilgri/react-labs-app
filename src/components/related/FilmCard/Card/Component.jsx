import React, { useEffect } from "react";
import PropTypes from "prop-types";

import FilmBox from "./FilmBox";
import TicketBox from "./TicketBox";
import { Wrapper, Pending } from "./styles";

const Component = ({
  filmId,
  fetchFilmByIdRequested,
  film,
  filmError,
  loading,
  user,
  makeOrderRequested,
  orderError,
  orderLoading,
  updateFilmRatingRequested,
}) => {
  const handleOrder = ({ chosenDate }) => {
    makeOrderRequested({
      filmId: film.id,
      screeningDate: chosenDate,
      userUid: user.uid,
    });
  };

  const handleRatingChange = (rating) => {
    updateFilmRatingRequested({ rating, filmId: film.id, userUid: user.uid });
  };

  useEffect(() => {
    fetchFilmByIdRequested(filmId);
  }, []);

  if (loading || filmError) {
    return <Pending>Loading</Pending>;
  }

  return (
    <Wrapper>
      <FilmBox
        title={film.title}
        description={film.description}
        image={film.image}
        ticketPrice={film.ticketPrice}
        tags={film.tags}
        user={user}
        rating={film.rate}
        onRatingChange={handleRatingChange}
      />
      <TicketBox
        screeningDates={film.screeningDates}
        user={user}
        onOrder={handleOrder}
        orderLoading={orderLoading}
        orderError={orderError}
      />
    </Wrapper>
  );
};

Component.defaultProps = {
  fetchFilmByIdRequested: () => {},
  makeOrderRequested: () => {},
  filmId: "",
  film: {},
  filmError: "",
  loading: false,
  user: null,
  orderError: "",
  orderLoading: false,
  updateFilmRatingRequested: () => {},
};

Component.propTypes = {
  filmId: PropTypes.string.isRequired,
  fetchFilmByIdRequested: PropTypes.func.isRequired,
  film: PropTypes.object,
  filmError: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  user: PropTypes.object,
  makeOrderRequested: PropTypes.func.isRequired,
  orderError: PropTypes.string,
  orderLoading: PropTypes.bool.isRequired,
  updateFilmRatingRequested: PropTypes.func.isRequired,
};

export default Component;
