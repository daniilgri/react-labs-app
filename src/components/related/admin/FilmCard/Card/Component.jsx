import React, { useEffect } from "react";
import PropTypes from "prop-types";

import FilmBox from "./FilmBox";
import TicketBox from "./TicketBox";
import { Wrapper, Pending } from "./styles";

const Component = ({ filmId, fetchFilmByIdAdminPanelRequested, film, filmError, loading }) => {
  useEffect(() => {
    fetchFilmByIdAdminPanelRequested(filmId);
  }, [fetchFilmByIdAdminPanelRequested, filmId]);

  if (loading || filmError || !film) {
    return <Pending>Loading</Pending>;
  }

  return (
    <Wrapper>
      <FilmBox film={film} />
      <TicketBox screeningDates={film.screeningDates} />
    </Wrapper>
  );
};

Component.defaultProps = {
  filmError: "",
  filmId: "",
  film: {
    title: "",
    description: "",
    tags: [],
    image: "",
    rates: [],
    screeningDates: [],
    ticketPrice: "",
  },
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchFilmByIdAdminPanelRequested: PropTypes.func.isRequired,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ),
  filmError: PropTypes.string,
  loading: PropTypes.bool.isRequired,
};

export default Component;
