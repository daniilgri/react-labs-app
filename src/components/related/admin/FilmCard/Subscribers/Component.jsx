import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import Card from "./Card";

import { Wrapper, Loading, CenterContainer, FetchButton } from "./styles";

const Component = ({
  filmId,
  fetchSubscribersInitialRequested,
  users,
  loading,
  error,
  allCount,
  count,
  fetchSubscribersNextRequested,
}) => {
  const handleGetMoreButtonOnClick = event => {
    event.preventDefault();

    if (count < allCount) {
      fetchSubscribersNextRequested({ filmId });
    }
  };

  useEffect(() => {
    fetchSubscribersInitialRequested({ filmId });
  }, []);

  return (
    <>
      <Wrapper>
        {users.length > 0 && users.map(item => <Card key={uuidv4()} user={item} />)}
      </Wrapper>
      {loading || error ? (
        <Loading>Loading</Loading>
      ) : (
        count < allCount && (
          <CenterContainer>
            <FetchButton onClick={handleGetMoreButtonOnClick}>Get more</FetchButton>
          </CenterContainer>
        )
      )}
    </>
  );
};

Component.defaultProps = {
  users: [],
  error: "",
  count: 0,
  allCount: 0,
  filmId: "",
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchSubscribersInitialRequested: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(PropTypes.object),
  error: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  count: PropTypes.number,
  allCount: PropTypes.number,
  fetchSubscribersNextRequested: PropTypes.func.isRequired,
};

export default Component;
