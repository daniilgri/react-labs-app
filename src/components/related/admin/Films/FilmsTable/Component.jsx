import React, { useEffect } from "react";
import PropTypes from "prop-types";
import Card from "./Card";
import { v4 as uuidv4 } from "uuid";

import { Wrapper, Loading } from "./styles";

const Component = ({
  loading,
  fetchFilmsAdminPanelInitialRequested,
  films,
  error,
  deleteFilmRequested,
}) => {
  useEffect(() => {
    fetchFilmsAdminPanelInitialRequested({ count: 25 });
  }, []);

  let content;

  if (loading || error) {
    content = <Loading>Loading</Loading>;
  } else {
    content =
      films.length > 0 ? (
        films.map((item) => (
          <Card key={uuidv4()} film={item} onDelete={deleteFilmRequested} />
        ))
      ) : (
        <Loading>Empty</Loading>
      );
  }

  return <Wrapper>{content}</Wrapper>;
};

Component.defaultProps = {
  films: [],
  loading: true,
  error: "",
  fetchFilmsAdminPanelInitialRequested: () => {},
  deleteFilmRequested: () => {},
};

Component.propTypes = {
  films: PropTypes.array,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  fetchFilmsAdminPanelInitialRequested: PropTypes.func.isRequired,
  deleteFilmRequested: PropTypes.func.isRequired,
};

export default Component;
