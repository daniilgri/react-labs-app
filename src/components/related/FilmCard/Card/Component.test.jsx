import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import Component from "./Component";

describe("FilmCard Card component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <Component
            filmId="filmId"
            fetchFilmByIdRequested={() => {}}
            film={{
              title: "",
              description: "",
              tags: [],
              image: "",
              rates: [],
              screeningDates: [],
              ticketPrice: "",
            }}
            loading={false}
            makeOrderRequested={() => {}}
            orderLoading={false}
            updateFilmRatingRequested={() => {}}
            loggedIn={false}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
