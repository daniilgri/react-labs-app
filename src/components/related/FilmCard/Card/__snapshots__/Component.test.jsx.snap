// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmCard Card component renders correctly 1`] = `
<div
  className="sc-bBXqnf cjhDUR"
>
  <div
    className="sc-hKgILt eJgxFi"
  >
    <div
      className="sc-eCssSg bTRJGu"
      src=""
    />
    <div
      className="sc-jSgupP jirZjF"
    >
      <ul
        className="sc-pFZIQ foKRaL"
      />
      <div
        className="sc-kEjbxe iUyqHI"
      >
        <h5
          className="sc-crrsfI dWZedv"
        >
          Rate 
          <span
            className="sc-iqHYGH fdsePL"
          >
            -
          </span>
        </h5>
      </div>
      <h4
        className="sc-fubCfw kOumzu"
      >
        
         $
      </h4>
      <h2
        className="sc-gKsewC dDbFlM"
      >
        
      </h2>
      <p
        className="sc-iBPRYJ kBXfFB"
      >
        
      </p>
    </div>
  </div>
  <div
    className="sc-dQppl eUOehi"
  >
    <h3
      className="sc-bqyKva xRQaq"
    >
      Ticketing
    </h3>
    <div
      className="sc-kstrdz cpvyEA"
    >
      <div
        className="sc-hBEYos bdYBED"
      >
        <h5
          className="sc-fodVxV hBIcmo"
        >
          Day
        </h5>
        <ul
          className="sc-fFubgz fTUZBb"
        />
      </div>
    </div>
    <div
      className="sc-fKFyDc kaLCkD"
    >
      <button
        className="sc-dIUggk bmuTPx"
        onClick={[Function]}
        type="button"
      >
        Order
      </button>
    </div>
  </div>
</div>
`;
